### PR TITLE
TMEDIA-675 - Button Component

### PR DESCRIPTION
**Pull request by matthewroach**
## Ticket

- [TMEDIA-675](https://arcpublishing.atlassian.net/browse/TMEDIA-675)

## Description

Added a `<Button>` component as per ticket, and added the following additions

* Update the Sass library to have a new mixin to reduce the need to specify all states in the `.scss` file using `component-with-states()`
* Added additional documentation around Accessibility for the component
* Added `classnames` module for handling `className`

## Acceptance Criteria

* A button can be a label only, icon only, label and icon, or label and dual icons
* Variations:
    * Primary (solid) button and reverse - Large, Medium, and Small sizes
    * Secondary (outline) button and reverse - Large, Medium, and Small sizes
    * Default button - defaults to default styling
* A button can render and act as a link if passed as href 
* A button can have a type -- button (default), submit, or reset
* States of Default, Hover, Focus, Disabled, Active

## Test Steps

1. Checkout branch - `git checkout TMEDIA-675-button`
2. Update dependencies - `npm i`
3. Run Storybook `npm run storybook`
4. Check out the button storybook documentation

### CODE REVIEW

File: src/components/button/index.jsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		style,
+		type,
+		...rest
+	} = props;
+
+	const defaultProps = {
+		...rest,
+		ref,
+		"aria-label": accessibilityLabel,
+		className: classnames({
```
Comment (JackHowa) [inline src/components/button/index.jsx]: not totally sure I understand how this lib is better than string interpolation. I had to lookup how this library works to understand these lines of code

File: src/components/button/index.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+<Meta title="Components/Button" component={Button} />
+
+# Button
+
+A Button represent a clickable element that is generally found within forms for submitting and resetting.
+
+Buttons are generally used for actions such as "Show More", "Save", "Cancel" - when a user activated the button involves an action to be taken.
+
+In the cases where a Button's appearance is needed but the intention is to take the user to another page use the `href` prop to have the visual appearance of your button but using an `a` link.
```
Comment (JackHowa) [inline src/components/button/index.stories.mdx:16]: nice explanation

**Review by JackHowa — COMMENTED** [review]
Looks good to me based on the storybook. Is there any documentation for the scss with state? 

Also, as a separate note, I found using the classname library to be less readable than interpolating the strings. I wasn't sure of the reason for using this library.

File: src/components/button/index.jsx (hunk 1 above)
Comment (matthewroach) [reply src/components/button/index.jsx]: With the amount of classNames we can have with the button, doing string interpolation looked very messy - essential with this library if the value is a truthy value it would add the key value to the class list

**Comment by matthewroach** [comment]
> Looks good to me based on the storybook. Is there any documentation for the scss with state?

We can look to add documentation for the Scss library mixins - not something we have publicly documented as of yet.

File: src/components/button/index.jsx (hunk 1 above)
Comment (matthewroach) [reply src/components/button/index.jsx]: @JackHowa - How would you accomplish it without a library?

File: src/components/button/index.jsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
+/* eslint-disable react/button-has-type */
+/*
+	Disabled ESLint rule as the type attribute is using oneOf propType with a default
+	value - whihch ESLint is unable to parse as acceptable when in the use case below
```
Comment (JackHowa) [inline src/components/button/index.jsx]:
```suggestion
	value - which ESLint is unable to parse as acceptable when in the use case below
```

Comment (matthewroach) [reply src/components/button/index.jsx]: Updated

File: src/components/button/index.jsx (hunk 1 above)
Comment (nschubach) [reply src/components/button/index.jsx]: as long as we are bringing this in, it could also be used for the "classname merge" that we have in all the components.  It's cleaner than the "append" method we have currently.

Comment (JackHowa) [reply src/components/button/index.jsx]: I get that interpolation can look messy. In terms of readability, and for new or newer devs, this seems more readable to me compared with the current text. 

From a testing perspective, this is also more reproducible. 




```jsx
	const defaultProps = {
		...rest,
		ref,
		"aria-label": accessibilityLabel,
		className: `${COMPONENT_CLASS_NAME}${href ? ` ${COMPONENT_CLASS_NAME}--link` : ""}${fullWidth ? ` ${COMPONENT_CLASS_NAME}--full-width` : ""}${size ? ` ${COMPONENT_CLASS_NAME}--${size}` : ""}${style ? ` ${COMPONENT_CLASS_NAME}--${style}` : ""}${className ? ` ${className}` : ""}`,
	};
```

One suggestion that would help me as reader is a comment above classname that might say what the library does. That might be helpful if I was looking at this with fresh eyes.

Comment (JackHowa) [reply src/components/button/index.jsx]: as a side note, I don't think that deduping passed-in classnames is a requirement, right? That doesn't cause issues with duplicated classnames nor would I expect devs to accidentally do this.

File: src/scss/root/_index.scss [hunk 4]
```diff
@@ -54,6 +54,18 @@ $tokens: () !default;
 	}
 }
 
+@mixin component-with-states($name, $states: []) {
```
Comment (nschubach) [inline src/scss/root/_index.scss:57]: This is interesting, but I wonder if it might be confusing or be abused for things that are not "state"

File: src/components/button/_index.scss [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	&:focus {
+		@include scss.component-properties("button-focus");
+	}
+
+	&[aria-disabled] {
+		cursor: not-allowed;
+
+		@include scss.component-with-states("button-disabled", [hover, active]);
```
Comment (nschubach) [inline src/components/button/_index.scss:27]: Are "hover" and "active" meant to be strings?

Comment (matthewroach) [reply src/components/button/_index.scss:27]: Not in Sass they don't have to be

File: src/components/button/index.jsx (hunk 1 above)
Comment (matthewroach) [reply src/components/button/index.jsx]: With Prettier installed that translates too

```jsx
		className: `${COMPONENT_CLASS_NAME}${href ? ` ${COMPONENT_CLASS_NAME}--link` : ""}${
			fullWidth ? ` ${COMPONENT_CLASS_NAME}--full-width` : ""
		}${size ? ` ${COMPONENT_CLASS_NAME}--${size}` : ""}${
			style ? ` ${COMPONENT_CLASS_NAME}--${style}` : ""
		}${className ? ` ${className}` : ""}`,
```

which is very hard to follow what's going on

Comment (nschubach) [reply src/components/button/index.jsx]: IMO, the `classnames` library is easier to read and I don't think it's too hard to see what's going on for someone new to it.

If you format it as this, prettier won't touch it:
```jsx
		className: `
			${COMPONENT_CLASS_NAME}
			${href ? `${COMPONENT_CLASS_NAME}--link` : ""}
			${fullWidth ? `${COMPONENT_CLASS_NAME}--full-width` : ""}
			${size ? `${COMPONENT_CLASS_NAME}--${size}` : ""}
			${style ? `${COMPONENT_CLASS_NAME}--${style}` : ""}
			${className ? `${className}` : ""}
		`,
```
...but it would also linebreak the classname property in the rendered output.

There is an alternative if we have an issue with the popular `classnames` that offers some additional features: https://www.npmjs.com/package/classd  (but I think `classnames` is probably fine)

I do see a lot of disfavor on-line toward using `classnames` though and I don't know why.

Comment (JackHowa) [reply src/components/button/index.jsx]: yeah I'm fine if this is not the way we go. one takeaway for me is that a comment above the lib would have been helpful for me to understand what was happening. Is that an ok compromise? @matthewroach 

Explicit, albeit messy (with prettier), would be preferable. As a bias, I care less if something is 'clean' or code-golf shortest solution. When I see a ternary, I know it's a bool check. Whereas, in that lib, I didn't know that the value was a bool check

File: src/components/button/_index.scss [hunk 6]
```diff
@@ -0,0 +1,64 @@
+@use "../../scss";
+
+.c-button {
+	-webkit-appearance: none;
```
Comment (nschubach) [inline src/components/button/_index.scss:4]: The `-webkit` prefix is probably not needed here (anymore).  https://developer.mozilla.org/en-US/docs/Web/CSS/appearance

Comment (matthewroach) [reply src/components/button/_index.scss:4]: This is for mobile safari - as it still changes the appearance of the button

File: src/components/button/index.jsx (hunk 1 above)
Comment (matthewroach) [reply src/components/button/index.jsx]: I am not sold on the comment above the library - as the more we use the library do we still add a comment every-time we use it? Feels like an overly cautious approach that I don't fully understand.

If we are happy with using the library I think we should update all components to use it - which would then ensure developers understand the library and get used to using it

File: src/scss/root/_index.scss (hunk 4 above)
Comment (nschubach) [reply src/scss/root/_index.scss:57]: I think this is probably fine as is... if anything comes up we can always revisit.

File: src/components/button/index.jsx (hunk 1 above)
Comment (nschubach) [reply src/components/button/index.jsx]: @JackHowa `classnames` can also be used like this since falsey values are ignored:
```
		className: classnames(
			COMPONENT_CLASS_NAME,
			href && `${COMPONENT_CLASS_NAME}--link`,
			fullWidth && `${COMPONENT_CLASS_NAME}--full-width`,
			size && `${COMPONENT_CLASS_NAME}--${size}`,
			style && `${COMPONENT_CLASS_NAME}--${style}`,
			className,
		),
```
Or  combination of the two methods:
```
		className: classnames(
			COMPONENT_CLASS_NAME,
			className,
			{
				[`${COMPONENT_CLASS_NAME}--link`]: href,
				[`${COMPONENT_CLASS_NAME}--full-width`]: fullWidth,
				[`${COMPONENT_CLASS_NAME}--${size}`]: size,
				[`${COMPONENT_CLASS_NAME}--${style}`]: style,
			}
		),
```

Comment (nschubach) [reply src/components/button/index.jsx]: also ran across this library that claims to be faster: https://www.npmjs.com/package/clsx

File: src/components/button/_index.scss (hunk 6 above)
Comment (nschubach) [reply src/components/button/_index.scss:4]: (edit to remove old screen cap)

https://caniuse.com/?search=appearance

Comment (nschubach) [reply src/components/button/_index.scss:4]: it seems that `appearance: "none"` is more supported...
image

File: src/components/button/_index.scss [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+	&:active {
+		@include scss.component-properties("button-active");
+	}
+
+	&:focus {
+		@include scss.component-properties("button-focus");
+	}
+
+	&[aria-disabled] {
```
Comment (ecarlisle) [inline src/components/button/_index.scss:24]: It's really good that we're using `aria-disabled` to ensure screen readers can see the button. Should we also have a reduced opacity in this selector like the effect of instead using the `disabled` attribute?  If not, are we relying on a token for this?

```
&[aria-disabled] {
  cursor: not-allowed;
  opacity: 0.5
}
```

Here's a reference I found for this:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-disabled

File: src/components/button/index.jsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+	if (href) {
+		return (
+			<a {...defaultProps} href={href}>
+				{buttonContents}
+			</a>
+		);
+	}
+	return (
+		<button {...defaultProps} type={type} aria-disabled={disabled}>
```
Comment (ecarlisle) [inline src/components/button/index.jsx:60]: Am I correct that `aria-disabled` allows a screen reader to see the button, but the button isn't disabled? Also, we shouldn't use `disabled` as that makes the button less inclusive. Will we need to write checks for this attribute into event handlers in other components, blocks as well as documenting that requirement?

I wrote this test HTML to check:
```
  <button class="foo" aria-disabled="true">Press Me</button>
  <script>
    document.querySelector('.foo').addEventListener('click', function(){
      alert('foo');
    });
  </script>
  ```
  
  In this case the event listener executes.

**Review by ecarlisle — COMMENTED** [review]
Wanted to confirm disabled button behavior. Thanks!

File: src/components/button/index.jsx (hunk 1 above)
Comment (ecarlisle) [reply src/components/button/index.jsx]: @nschubach : Great find! Here's an Npm Trends comparison:
[npmtrends.com/clsx-vs-classnames](https://www.npmjs.com/package/clsx)

The size difference is marginal:
https://bundlephobia.com/package/classnames@2.3.1
https://bundlephobia.com/package/clsx@1.1.1

But the readme for `clsx` shows benchmarks between the two:
https://github.com/lukeed/clsx/tree/master/bench#benchmarks

Comment (matthewroach) [reply src/components/button/index.jsx]: I looked at clsx but their document was very limited - and I wasn't sure if they followed the `classnames` functionality - if we are using a third party library I want the documentation to be reliable

File: src/components/button/index.jsx (hunk 8 above)
Comment (matthewroach) [reply src/components/button/index.jsx:60]: Correct as per the documentation for the component - the `disabled` attribute isn't the best attribute to use for disabling a button as assistive technologies can not always get access to a disabled button

> The prop `disabled` uses the `aria-disabled` attribute on a `button` to provide assistive technologies the ability to still access the `button` in the tab order, but provide context to the users of assistive technologies.
When using the `disabled` property any events will need to account for preventing the event completing.

File: src/components/button/_index.scss (hunk 7 above)
Comment (matthewroach) [reply src/components/button/_index.scss:24]: We should avoid putting too much style in to the base component - I'd like to rely on the token system here for appearance

File: src/components/button/_index.scss (hunk 6 above)
Comment (matthewroach) [reply src/components/button/_index.scss:4]: `appearance: none` doesn't work on my iPad - it needs `-webkit-appearance: none`

File: src/components/button/index.jsx (hunk 1 above)
Comment (JackHowa) [reply src/components/button/index.jsx]: > I am not sold on the comment above the library - as the more we use the library do we still add a comment every-time we use it? Feels like an overly cautious approach that I don't fully understand.

Against adding one comment in multiple places? Adding comments to me helps comprension wherever I see something confusing, even more so if repeated. 

If opposed to a one-line comment "// the value of the key-value pair is a boolean check to determine whether or not to show the key as a class"?, is there another way we can be explicit? For example, what about explicitly casting the values as boolean because that's how they're being used:

```
[`${COMPONENT_CLASS_NAME}--full-width`]: Boolean(fullWidth),
```

Overall, my initial impression to this code was surprise that we were both passing a boolean as well as a string as a value into a library I had never seen before. In my mind, this is a minor readability suggestion. I haven't looked at other libraries doing conditional string interpolation.

Comment (matthewroach) [reply src/components/button/index.jsx]: Update the implementation to remove the need for the `classnames` library - having looked at @nschubach example about it occurred to me we can do the same logic with a basic array

**Review by JackHowa — APPROVED** [review]
Looks good! Thanks for discussing the class name

File: src/components/button/index.jsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+		className: [
+			COMPONENT_CLASS_NAME,
+			href && `${COMPONENT_CLASS_NAME}--link`,
+			fullWidth && `${COMPONENT_CLASS_NAME}--full-width`,
+			size && `${COMPONENT_CLASS_NAME}--${size}`,
+			variant && `${COMPONENT_CLASS_NAME}--${variant}`,
+			className,
+		]
+			.filter((i) => i)
+			.join("  "),
```
Comment (JackHowa) [inline src/components/button/index.jsx]:
```suggestion
			.join(" "),
```

we want one space not two between classes, right?

Comment (JackHowa) [reply src/components/button/index.jsx]: Doesn't wildly make a difference but just noticed it

**Review by nschubach — APPROVED** [review]

**Review by JackHowa — APPROVED** [review]
Good work!
